server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransaction.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.tserver.tablet;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.accumulo.core.metadata.StoredTabletFile;
+
+public abstract class DatafileTransaction {
+
+  protected final long ts = System.currentTimeMillis();
```

Comment:
Is it possible to have two transactions on a tablet in the same millisecond?

Reply:
I share this concern, but also System.currentTimeMillis() is not necessarily ordered. It can go backwards. This is the user's clock. Lots of things can change it, including NTP, the user, another process, etc. There is no guarantee this is monotonically increasing, nor is it guaranteed that it is increasing at all for transactions that happen later.

Reply:
These times are for diagnostic purposes only.  I can change this to nano time instead.  I know System.currentTimeMillis() is not precise but there was not need for precision here.

Reply:
Actually nanoTime does not help me here either.  What exactly are you suggesting I do here?

Reply:
If you're only using it for diagnostic / informational purposes, then it's fine. Just can't use it as a unique key for a transaction or for ordering transactions, or for computing time differences between two transactions. Just informational should be fine.